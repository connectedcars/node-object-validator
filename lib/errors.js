function ValidationError(message, context = {}) {
  this.name = 'ValidationError'
  this.message = message
  if (context.key) {
    this.key = context.key
  }
  if (context.val) {
    this.val = context.val
  }
}
ValidationError.prototype = new Error()

module.exports = {
  ValidationError
}
