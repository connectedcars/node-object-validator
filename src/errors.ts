export interface ValidationErrorContext<T> {
  key: string
  value: T
}

export class ValidationError<T> extends Error {
  public key?: string
  public value?: T
  public constructor(message: string, context?: ValidationErrorContext<T>) {
    if (context) {
      message = `Field '${context.key}' ${message.charAt(0).toLowerCase() + message.slice(1)}`
    }
    super(message)
    this.key = context?.key
    this.value = context?.value
  }
}

export class RequiredError extends ValidationError<string> {}
export class NotStringError extends ValidationError<string> {}
export class WrongLengthError extends ValidationError<string> {}
export class NotDateError extends ValidationError<string> {}
export class NotRfc3339Error extends ValidationError<string> {}
export class NotExactStringError extends ValidationError<string> {}
export class NotFloatError extends ValidationError<string> {}
export class NotIntegerError extends ValidationError<string> {}
export class NotArrayError extends ValidationError<string> {}
export class NotObjectError extends ValidationError<string> {}
export class DoesNotMatchRegexError extends ValidationError<string> {}
export class OutOfRangeError extends ValidationError<string> {}
