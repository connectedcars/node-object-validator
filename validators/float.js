const { ValidationError } = require('../lib/errors')

function Float(min = Number.MIN_SAFE_INTEGER, max = Number.MAX_SAFE_INTEGER, required = true) {
  return {
    type: 'Float',
    validate: function(key, val) {
      if (val == null) {
        if (required) {
          return new ValidationError(`Field \`${key}\` (${this.type}) is required`)
        } else {
          return
        }
      }
      if (typeof val !== 'number' || isNaN(val) || !isFinite(val)) {
        return new ValidationError(`Field \`${key}\` (${this.type}) must be a float (received "${val}")`)
      }
      if (val < min) {
        return new ValidationError(`Field \`${key}\` (${this.type}) must at least be ${min} (received "${val}")`)
      }
      if (val > max) {
        return new ValidationError(`Field \`${key}\` (${this.type}) must at most be ${max} (received "${val}")`)
      }
    }
  }
}

module.exports = Float
