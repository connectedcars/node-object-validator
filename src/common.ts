import { RequiredFail, ValidationFailure, ValidationsError } from './errors'

// https://stackoverflow.com/questions/51651499/typescript-what-is-a-naked-type-parameter
// https://2ality.com/2019/07/testing-static-types.html
// Wrapping the types in an tuple force a specific type instead of allow any in the union
export type AssertEqual<T, Expected> = [T, Expected] extends [Expected, T] ? true : false

export type ReturnEqual<T, C> = [T, C] extends [C, T] ? C : never

// https://github.com/microsoft/TypeScript/issues/41746
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type BaseObject = Record<string, any>

// TODO: Give better name
export function isValidType<T>(value: unknown, errors: ValidationFailure[]): value is T {
  return errors.length === 0
}

export type CodeGenResult = [{ [key: string]: unknown }, string[], string[]]

export type ValidatorBaseOptions = {
  optimize?: boolean
  required?: boolean
  nullable?: boolean
  earlyFail?: boolean
}

/**
 * @typedef ValidatorOptions
 * @property {boolean} [earlyFail=false] Stop validation on first failure (default: false)
 * @property {boolean} [optimize=true] Generate an optimized function for doing the validation (default: true)
 */
export interface ValidatorOptions {
  earlyFail?: boolean
  optimize?: boolean
}

export interface ValidateOptions {
  earlyFail?: boolean
  optimized?: boolean
}

export interface ValidatorExportOptions {
  language?: string
  types?: boolean
}

export function isValidator(value: unknown): value is ValidatorBase {
  if (value instanceof ValidatorBase) {
    return true
  }
  return false
}

export function generateOptionsString(options: ValidatorBaseOptions, defaults: Required<ValidatorBaseOptions>): string {
  const selectedOptions: string[] = []
  if (options.required !== undefined && options.required !== defaults.required) {
    selectedOptions.push(`required: ${options.required}`)
  }
  if (options.nullable !== undefined && options.nullable !== defaults.nullable) {
    selectedOptions.push(`nullable: ${options.nullable}`)
  }
  if (options.earlyFail !== undefined && options.earlyFail !== defaults.earlyFail) {
    selectedOptions.push(`earlyFail: ${options.earlyFail}`)
  }
  if (options.optimize !== undefined && options.optimize !== defaults.optimize) {
    selectedOptions.push(`optimize: ${options.optimize}`)
  }
  return selectedOptions.length > 0 ? `{ ${selectedOptions.join(', ')} }` : ''
}

export abstract class ValidatorBase<T = unknown> {
  public tsType!: T
  public required: boolean
  public nullable: boolean
  public earlyFail: boolean

  protected codeGenId = 1
  protected optimizedValidate: ((value: unknown, context?: string) => ValidationFailure[]) | null
  protected optionsString: string

  public constructor(options?: ValidatorBaseOptions) {
    const defaults = { required: true, nullable: false, earlyFail: false, optimize: true }
    const mergedOptions = { ...defaults, ...options }

    this.optionsString = options ? generateOptionsString(options, defaults) : ''

    this.required = mergedOptions.required
    this.earlyFail = mergedOptions.earlyFail
    this.nullable = mergedOptions.nullable
    this.optimizedValidate = null
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public AssertType<C extends T, X extends AssertEqual<T, C>>(): void {
    // Only used for type assertions
  }

  public isValid<C extends T>(obj: unknown): obj is ReturnEqual<T, C> {
    const errors = this.validate(obj)
    return errors.length === 0
  }

  public isType<C extends T>(obj: unknown, errors: ValidationFailure[]): obj is ReturnEqual<T, C> {
    return errors.length === 0
  }

  public cast<C extends T>(obj: unknown): ReturnEqual<T, C> {
    const errors = this.validate(obj)
    if (this.isType<C>(obj, errors)) {
      return obj
    } else {
      throw new ValidationsError('One of more validations failed', errors)
    }
  }

  public validate(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[] {
    if (options?.optimized !== false && this.optimizedValidate !== null) {
      return this.optimizedValidate(value, context)
    }
    if (this.nullable && value === null) {
      return []
    }
    if (value === undefined) {
      return this.required ? [new RequiredFail(`Is required`, value, context)] : []
    }
    return this.validateValue(value, context, { earlyFail: this.earlyFail, optimized: false, ...options })
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, { key: (context ? \`\${context}['${context}']\` : \`${context}\`) }` : ', context'
    const validatorName = `validator${id()}`
    const declarations = [`const ${validatorName} = ${validatorRef}`]
    // prettier-ignore
    const code: string[] = [
      ...this.nullCheckWrap([
        `  errors.push(...${validatorName}.validateValue(${valueRef}${contextStr}))`,
      ],valueRef, contextStr),
      ...(this.earlyFail || earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),
    ]
    return [
      {
        RequiredFail: RequiredFail
      },
      declarations,
      code
    ]
  }

  protected nullCheckWrap(code: string[], valueRef: string, contextStr: string, skip = false): string[] {
    if (skip) {
      return code
    }

    const result: string[] = []
    if (this.required) {
      if (this.nullable) {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(`  if (${valueRef} !== null) {`)
        result.push(...code)
        result.push(`  }`)
        result.push(`} else {`)
        result.push(`  errors.push(new RequiredFail(\`Is required\`, ${valueRef}${contextStr}))`)
        result.push(`}`)
      } else {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(...code)
        result.push(`} else {`)
        result.push(`  errors.push(new RequiredFail(\`Is required\`, ${valueRef}${contextStr}))`)
        result.push(`}`)
      }
    } else {
      if (this.nullable) {
        result.push(`if (${valueRef} != null) {`)
        result.push(...code)
        result.push(`}`)
      } else {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(...code)
        result.push(`}`)
      }
    }
    return result
  }

  protected optimize(schema?: unknown): void {
    const [imports, declarations, code] = this.codeGen(
      'value',
      'schema',
      () => {
        return this.codeGenId++
      },
      undefined,
      this.earlyFail
    )
    const functionBody = [
      ...Object.keys(imports).map(i => `const ${i} = imports['${i}']`),
      ...declarations,
      `return (value, context) => {`,
      `  const generatedFunction = true`,
      `  const errors = []`,
      ...code.map(l => `  ${l}`),
      `  return errors`,
      `}`
    ].join('\n')

    try {
      const functionGenerator = new Function('imports', 'schema', functionBody)
      const validateFunction = functionGenerator(imports, {
        schema,
        validateValue: this.validateValue.bind(this)
      })
      this.optimizedValidate = validateFunction
    } catch (e) {
      throw new Error(`Failed to compile optimized function(${e.message}):\n${functionBody}`)
    }
  }

  public abstract toString(options?: ValidatorExportOptions): string

  protected abstract validateValue(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[]
}
