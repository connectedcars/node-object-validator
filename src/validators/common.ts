import { ValidationErrorContext } from '../errors'

export type Schema = {
  [key: string]: Validator
}

export interface Validator {
  schema?: Schema
  validate(value: unknown, context?: ValidationErrorContext): Error | null
}
