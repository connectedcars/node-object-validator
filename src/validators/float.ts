import { ValidatorBase } from '../common'
import { NotFloatError, OutOfRangeError, RequiredError, ValidationErrorContext } from '../errors'

export function validateFloat(
  value: unknown,
  min = Number.MIN_SAFE_INTEGER,
  max = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): Error[] {
  if (typeof value !== 'number' || isNaN(value) || !isFinite(value)) {
    return [new NotFloatError(`Must be a float (received "${value}")`, context)]
  }
  if (value < min || value > max) {
    return [new OutOfRangeError(`Must be between ${min} and ${max} (received "${value}")`, context)]
  }
  return []
}

export class RequiredFloat extends ValidatorBase<number> {
  private type: 'RequiredFloat' = 'RequiredFloat'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    super()
    this.min = min
    this.max = max
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] {
    if (value == null) {
      return [new RequiredError(`Is required`, context)]
    }
    return validateFloat(value, this.min, this.max, context)
  }
}

export class OptionalFloat extends ValidatorBase<number | undefined | null> {
  private type: 'OptionalFloat' = 'OptionalFloat'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    super()
    this.min = min
    this.max = max
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] {
    if (value == null) {
      return []
    }
    return validateFloat(value, this.min, this.max, context)
  }
}

export function Float(min: number, max: number, required?: false): OptionalFloat
export function Float(min: number, max: number, required: true): RequiredFloat
export function Float(min = 0, max = Number.MAX_SAFE_INTEGER, required = false): OptionalFloat | RequiredFloat {
  return required ? new RequiredFloat(min, max) : new OptionalFloat(min, max)
}
