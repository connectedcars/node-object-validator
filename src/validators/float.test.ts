import { AssertEqual } from '../common'
import { NotFloatFail, OutOfRangeFail, RequiredFail } from '../errors'
import { isFloat, NullableFloat, OptionalFloat, OptionalNullableFloat, RequiredFloat, validateFloat } from './float'

describe('Float', () => {
  describe('validateFloat', () => {
    it('requires value to be a float', () => {
      expect(validateFloat(0.0001)).toStrictEqual([])
    })
  })
  describe('isFloat', () => {
    it('should cast value to float', () => {
      const value = 0.0001 as unknown
      if (isFloat(value)) {
        expect(true as AssertEqual<typeof value, number>).toEqual(true)
      } else {
        fail('did not validate but should')
      }
    })

    it('should fail validation', () => {
      const value = 'string' as unknown
      expect(isFloat(value)).toEqual(false)
    })
  })

  describe('RequiredFloat', () => {
    it('should return an function body', () => {
      const validator = new RequiredFloat(0, 10, { optimize: false })
      expect(validator.codeGen('value1', 'validator1')).toMatchSnapshot()
    })

    it('should export types', () => {
      const validator = new RequiredFloat(0, 10, { optimize: false })
      const code = validator.toString({ types: true })
      expect(code).toEqual('number')
    })
  })
})

describe.each([false, true])('Float (optimize: %s)', optimize => {
  describe('RequiredFloat', () => {
    it('should generate validation code and give same result', () => {
      const validator = new RequiredFloat(1, 2, { optimize })
      if (optimize) {
        expect(validator['optimizedValidate']).not.toBeNull()
      } else {
        expect(validator['optimizedValidate']).toBeNull()
      }
      const errors = validator.validate(1.5)
      expect(errors).toEqual([])
    })

    it('should export validator code with options', () => {
      const validator = new RequiredFloat(1, 2, { optimize })
      const code = validator.toString()
      if (optimize) {
        expect(code).toEqual('new RequiredFloat(1, 2)')
      } else {
        expect(code).toEqual('new RequiredFloat(1, 2, { optimize: false })')
      }
    })

    it('accepts valid values', () => {
      const validator = new RequiredFloat(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER, { optimize })
      expect(validator.validate(0.0001)).toStrictEqual([])
      expect(validator.validate(1)).toStrictEqual([])
      expect(validator.validate(1.25)).toStrictEqual([])
      expect(validator.validate(123)).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, number>).toEqual(true)
    })

    it('rejects invalid values', () => {
      const validator = new RequiredFloat(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER, { optimize })
      expect(validator.validate('1')).toStrictEqual([new NotFloatFail('Must be a float', '1')])
      expect(validator.validate('')).toStrictEqual([new NotFloatFail('Must be a float', '')])
      expect(validator.validate({})).toStrictEqual([new NotFloatFail('Must be a float', {})])
      expect(validator.validate([])).toStrictEqual([new NotFloatFail('Must be a float', [])])
      expect(validator.validate(true)).toStrictEqual([new NotFloatFail('Must be a float', true)])
      expect(validator.validate(false)).toStrictEqual([new NotFloatFail('Must be a float', false)])
      expect(validator.validate(null)).toStrictEqual([new NotFloatFail('Must be a float', null)])
      expect(true as AssertEqual<typeof validator.tsType, number>).toEqual(true)
    })

    it('requires min value', () => {
      const validator = new RequiredFloat(0.5, 500, { optimize })
      expect(validator.validate(-0.1)).toStrictEqual([new OutOfRangeFail('Must be between 0.5 and 500', -0.1)])
      expect(validator.validate(0)).toStrictEqual([new OutOfRangeFail('Must be between 0.5 and 500', 0)])
      expect(validator.validate(0.1)).toStrictEqual([new OutOfRangeFail('Must be between 0.5 and 500', 0.1)])
      expect(validator.validate(0.2)).toStrictEqual([new OutOfRangeFail('Must be between 0.5 and 500', 0.2)])
      expect(validator.validate(0.3)).toStrictEqual([new OutOfRangeFail('Must be between 0.5 and 500', 0.3)])
      expect(validator.validate(0.4)).toStrictEqual([new OutOfRangeFail('Must be between 0.5 and 500', 0.4)])
      expect(validator.validate(0.49999999)).toStrictEqual([
        new OutOfRangeFail('Must be between 0.5 and 500', 0.49999999)
      ])
      expect(validator.validate(0.5)).toStrictEqual([])
      expect(validator.validate(0.6)).toStrictEqual([])
      expect(validator.validate(123.456)).toStrictEqual([])
    })

    it('requires max value', () => {
      const validator = new RequiredFloat(-500, 0.5, { optimize })
      expect(validator.validate(-0.1)).toStrictEqual([])
      expect(validator.validate(0)).toStrictEqual([])
      expect(validator.validate(0.1)).toStrictEqual([])
      expect(validator.validate(0.2)).toStrictEqual([])
      expect(validator.validate(0.3)).toStrictEqual([])
      expect(validator.validate(0.4)).toStrictEqual([])
      expect(validator.validate(0.5)).toStrictEqual([])
      expect(validator.validate(0.500000001)).toStrictEqual([
        new OutOfRangeFail('Must be between -500 and 0.5', 0.500000001)
      ])
      expect(validator.validate(0.6)).toStrictEqual([new OutOfRangeFail('Must be between -500 and 0.5', 0.6)])
      expect(validator.validate(0.7)).toStrictEqual([new OutOfRangeFail('Must be between -500 and 0.5', 0.7)])
    })

    it('rejects undefined', () => {
      const validator = new RequiredFloat(0, Number.MAX_SAFE_INTEGER, { optimize })
      expect(validator.validate(undefined)).toStrictEqual([new RequiredFail('Is required', undefined)])
    })

    it('requires value to show correct context on error', () => {
      const validator = new RequiredFloat(0, Number.MAX_SAFE_INTEGER, { optimize })
      expect(validator.validate('', 'float').map(e => e.toString())).toStrictEqual([
        `NotFloatFail: Field 'float' must be a float (received "")`
      ])
    })
  })

  describe('OptionalFloat', () => {
    it('accepts empty value', () => {
      const validator = new OptionalFloat(0, Number.MAX_SAFE_INTEGER, { optimize })
      expect(validator.validate(1.0)).toStrictEqual([])
      expect(validator.validate(undefined)).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, number | undefined>).toEqual(true)
    })
  })

  describe('NullableFloat', () => {
    it('accepts empty value', () => {
      const validator = new NullableFloat(0, Number.MAX_SAFE_INTEGER, { optimize })
      expect(validator.validate(1.0)).toStrictEqual([])
      expect(validator.validate(null)).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, number | null>).toEqual(true)
    })
  })

  describe('OptionalNullableFloat', () => {
    it('accepts empty value', () => {
      const validator = new OptionalNullableFloat(0, Number.MAX_SAFE_INTEGER, { optimize })
      expect(validator.validate(1.0)).toStrictEqual([])
      expect(validator.validate(null)).toStrictEqual([])
      expect(validator.validate(undefined)).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, number | null | undefined>).toEqual(true)
    })
  })
})
