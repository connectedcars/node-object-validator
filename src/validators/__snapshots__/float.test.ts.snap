// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Float RequiredFloat should return a function body 1`] = `
Array [
  Object {
    "NotFloatFail": [Function],
    "OutOfRangeFail": [Function],
    "RequiredFail": [Function],
  },
  Array [],
  Array [
    "const value1 = value1",
    "if (value1 !== undefined) {",
    "  if (typeof value1 === 'number' && !isNaN(value1 ) && isFinite(value1)) {",
    "    if (value1 < 0 || value1 > 10) {",
    "      errors.push(new OutOfRangeFail(\`Must be between 0 and 10\`, value1, context))",
    "    }",
    "  } else {",
    "    errors.push(new NotFloatFail(\`Must be a float\`, value1, context))",
    "  }",
    "} else {",
    "  errors.push(new RequiredFail(\`Is required\`, value1, context))",
    "}",
  ],
]
`;
